esp-hal/src/soc/mod.rs

```diff
@@ -246,3 +246,61 @@ fn setup_stack_guard() {
         ));
     }
 }
+
+#[cfg(all(feature = "rt", riscv, stack_guard_monitoring))]
+pub(crate) fn enable_stack_guard_monitoring() {
+    if !crate::debugger::debugger_connected() {
+        unsafe {
+            unsafe extern "C" {
+                static mut __stack_chk_guard: u32;
+            }
+
+            let guard_addr = core::ptr::addr_of_mut!(__stack_chk_guard) as *mut _ as u32;
+            assert!(guard_addr.is_multiple_of(4));
+
+            let id = 0; // breakpoint 0
+            let addr = (guard_addr & !0b11) | 0b01;
+
+            let tdata = (1 << 3) | (1 << 6) | (1 << 1); // bits: 0 = load, 1 = store, 6 = m-mode, 3 = u-mode
+            let tcontrol = 1 << 3; // M-mode trigger
+            core::arch::asm!(
+                "
+                csrw 0x7a0, {id} // tselect
+                csrrs {tcontrol}, 0x7a5, {tcontrol} // tcontrol
+                csrrs {tdata}, 0x7a1, {tdata} // tdata1
+                csrw 0x7a2, {addr} // tdata2
+                ", id = in(reg) id,
+                addr = in(reg) addr,
+                tdata = in(reg) tdata,
+                tcontrol = in(reg) tcontrol,
+            );
+        }
+    }
+}
+
+#[cfg(all(feature = "rt", xtensa, stack_guard_monitoring))]
+pub(crate) fn enable_stack_guard_monitoring() {
+    if !crate::debugger::debugger_connected() {
+        unsafe {
+            unsafe extern "C" {
+                static mut __stack_chk_guard: u32;
+            }
+
+            let guard_addr = core::ptr::addr_of_mut!(__stack_chk_guard) as *mut _ as u32;
+
+            assert!(guard_addr.is_multiple_of(4));
+            let addr = guard_addr & !0b11;
+
+            let dbreakc = 0b11 | (1 << 31); // bit 31 = STORE
+
+            core::arch::asm!(
+            "
+            wsr {addr}, 144
+            wsr {dbreakc}, 160
```

Comment:
Can we use `wsr.dbreakc0 x` and `wsr.dbreaka y` syntax instead?

Reply:
Also, what enables the CPU interrupt? Since a few PRs ago, level 6 is not unmasked by default.

Reply:
in that case, could you add the names in a comment? looking up the numbers in the ISA doc is a pain

Reply:
Ah `Level6.mask()` = 0 - so it's not really maskable